saleor/order/actions.py

```diff
@@ -388,3 +399,191 @@ def create_fulfillments(
+            fulfillment_lines_to_update.append(refunded_line)
+        lines_to_deallocate_stock.append((line_to_refund, unfulfilled_to_refund))
+        all_refunded_lines[line_to_refund.id] = (
+            unfulfilled_to_refund,
+            line_to_refund,
+        )
+    return refund_amount
+
+
+def _prepare_refund_fulfillment_for_fulfillment_lines(
+    fulfillment_lines_to_refund,
+    fulfillment_lines_quantities_to_refund,
+    already_refunded_lines,
+    refunded_fulfillment,
+    fulfillment_lines_to_create,
+    fulfillment_lines_to_update,
+    empty_fulfillment_lines_to_delete,
+    all_refunded_lines,
+):
+    refund_amount = Decimal(0)
+    order_lines_with_fulfillment = OrderLine.objects.in_bulk(
+        [line.order_line_id for line in fulfillment_lines_to_refund]
+    )
+    for fulfillment_line, quantity_to_refund in zip(
+        fulfillment_lines_to_refund, fulfillment_lines_quantities_to_refund
+    ):
+        refunded_line = _get_fulfillment_line_if_exists(
+            already_refunded_lines,
+            fulfillment_line.order_line_id,
+            fulfillment_line.stock_id,
+        )
+        fulfillment_line_existed = True
+        if not refunded_line:
+            fulfillment_line_existed = False
+            refunded_line = FulfillmentLine(
+                fulfillment=refunded_fulfillment,
+                order_line_id=fulfillment_line.order_line_id,
+                stock_id=fulfillment_line.stock_id,
+                quantity=0,
+            )
+        fulfilled_to_refund = min(fulfillment_line.quantity, quantity_to_refund)
+        quantity_to_refund -= fulfilled_to_refund
+        refunded_line.quantity += fulfilled_to_refund
+        fulfillment_line.quantity -= fulfilled_to_refund
+        order_line: OrderLine = order_lines_with_fulfillment.get(  # type: ignore
+            fulfillment_line.order_line_id
+        )
+        refund_amount += order_line.unit_price_gross_amount * fulfilled_to_refund
+        if fulfillment_line.quantity == 0:
+            empty_fulfillment_lines_to_delete.append(fulfillment_line)
+        else:
+            fulfillment_lines_to_update.append(fulfillment_line)
+        if refunded_line.quantity > 0 and not fulfillment_line_existed:
+            fulfillment_lines_to_create.append(refunded_line)
+        elif fulfillment_line_existed:
+            fulfillment_lines_to_update.append(refunded_line)
+
+        data_from_all_refunded_lines = all_refunded_lines.get(order_line.id)
+        if data_from_all_refunded_lines:
+            quantity, line = data_from_all_refunded_lines
+            quantity += fulfilled_to_refund
+            all_refunded_lines[order_line.id] = (quantity, line)
+        else:
+            all_refunded_lines[order_line.id] = (
+                fulfilled_to_refund,
+                order_line,
+            )  # type: ignore
+    return refund_amount
+
+
+def create_refund_fulfillment(
+    requester: Optional["User"],
+    order,
+    payment,
```

Comment:
Would be useful to have these params typed.

Reply:
These three functions are a bit hard to read. If possible I would try to add types to params where possible, docstrings, maybe some line comments to explain what is going on, or empty lines to separate pieces of related logic.

Reply:
I've cleaned it up. Now it should be better.